Create get task API that falls back to the .tasks index

nik9000 opened the pull request:
This was edited from its original form. The original is below for posterity:

## Rewrite
This adds a get task API that supports `GET /_tasks/${taskId}` and removes that responsibility from the list tasks API. The get task API supports wait_for_complation just as the list tasks API does but doesn't support any of the list task API's filters. In exchange, it supports falling back to the `.results` index when the task isn't running any more. Like any good GET API it 404s when it doesn't find the task.

Then we change reindex, update-by-query, and delete-by-query to persist the task result when `wait_for_completion=false`. The leads to the neat behavior that, once you start a reindex with `wait_for_completion=false`, you can fetch the result of the task by using the get task API and see the result when it has finished.


## Original 
This changes the task list API so that, for requests that look like `GET /_tasks/${taskId}`, it'll first look at the running tasks on the node and, if the task isn't found, perform a GetRequest on `.results/result/${taskId}`.

Then we change reindex, update-by-query, and delete-by-query to persist the task result when `wait_for_completion=false`.

The leads to the neat behavior that, once you start a reindex with `wait_for_completion=false`, you can fetch the result of the task by a task list request and you'll get the result, even if the task has finished.

nik9000 commented:
@imotov this is probably for you.

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksRequest.java:
```diff
@@ -63,6 +66,16 @@ public ListTasksRequest setWaitForCompletion(boolean waitForCompletion) {
+                && isEmpty(this.getNodesIds())
+                && isEmpty(getActions());
+    }
+
```
It feels like the single task mode significantly diverged from the list task mode in terms of parameters that we support, functionality that is provided and code that is getting executed in one mode vs another. Handling both in a single action becomes more and more hackish. It might be time to extract the single task mode into a separate `TransportGetTaskAction` with its own request and response. What do you think?

nik9000 replied:
Yeah, I think I'll do that.

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/TransportListTasksAction.java:
```diff
@@ -110,4 +150,215 @@ protected void processTasks(ListTasksRequest request, Consumer<Task> operation)
+            if (e instanceof NoSuchNodeException) {
+                startGet();
+            }
+            delegate.onFailure(e);
```
That looks like a race condition. I think it's also an important enough use case that is might be worth adding a test for.

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/TransportListTasksAction.java:
```diff
@@ -110,4 +150,215 @@ protected void processTasks(ListTasksRequest request, Consumer<Task> operation)
+                            // Didn't find the task either way so just return the empty response
+                            delegate.onResponse(new ListTasksResponse(emptyList(), emptyList(), emptyList()));
+                        }
+                    } catch (Exception e) {
```
Should this be Throwable?

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/TransportListTasksAction.java:
```diff
@@ -110,4 +150,215 @@ protected void processTasks(ListTasksRequest request, Consumer<Task> operation)
+    /**
+     * Builds TaskInfo from persisted copy in the task result index.
+     */
+    public static class TaskInfoBuilder {
```
That should probably go to TaskInfo, especially parser that should be definitely somewhere close to the corresponding toXContent method that generates this json.

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/TransportListTasksAction.java:
```diff
@@ -110,4 +150,215 @@ protected void processTasks(ListTasksRequest request, Consumer<Task> operation)
+        public static final ObjectParser<TaskInfoBuilder, ParseFieldMatcherSupplier> PARSER = new ObjectParser<>("task_result",
+                TaskInfoBuilder::new);
+        static {
+            // Note for the future: this has to be backwards compatible with all changes to the task persistence format
```
It might be a good idea to store a version of elasticsearch that generated the task somewhere on the top level. So, we could filter out incompatible versions in the future and don't have to support old persistence formats forever.

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/TransportListTasksAction.java:
```diff
@@ -110,4 +150,215 @@ protected void processTasks(ListTasksRequest request, Consumer<Task> operation)
+            DiscoveryNode node = new DiscoveryNode(null, nodeId, nodeId, nodeId, DummyTransportAddress.INSTANCE, emptyMap(), emptySet(),
+                    Version.CURRENT);
+            Task.Status status = new PersistedStatus(originalStatus, error, result);
+            return new TaskInfo(node, id, type, action, description, status, startTime, runningTimeNanos, false, parentTaskId);
```
We should make TaskInfo final then.

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/TransportListTasksAction.java:
```diff
@@ -110,4 +150,215 @@ protected void processTasks(ListTasksRequest request, Consumer<Task> operation)
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            if (originalStatus != null) {
+                XContentHelper.writeRawField("original_status", originalStatus, builder, params);
```
I am confused here. How is this different from the `status`?

imotov commented:
@nik9000 left a few comments.

nik9000 commented:
> @nik9000 left a few comments.

Sorry this is taking so long - mostly through the comment and I'll push something soon.

nik9000 commented:
@imotov I rebased and started on the get task API. Like the list tasks API it can wait for completion, but unlike the list tasks API it also falls back to the results index if the task isn't found. I didn't make that a disable-able thing.

nik9000 commented:
Note to self: make sure that we have a sufficient test case for the get task API's json output. This probably means expanding one reindex's tests because it is the only thing where we can be sure that the response came back without the test plugin.

nik9000 commented:
@imotov this is ready for another round I think. I have to go and fix up reindex's rest tests but it the stuff in core is ready.

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java:
```diff
@@ -161,33 +154,38 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                DiscoveryNode node = discoveryNodes.get(entry.getKey());
+                builder.startObject(entry.getKey());
+                if (node != null) {
+                    // If the node is no longer part of the cluster, oh well, we'll just skip it's useful information.
```
Hmm, not sure how I feel about that.

nik9000 replied:
I don't feel too bad about it because it should be super infrequent. I did it to make working with the TaskInfo in all the cases where you don't need the node's full information easier. And to shrink the wire format, but that wasn't that important to me.

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java:
```diff
@@ -161,33 +154,38 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
                 builder.startObject("tasks");
                 for(TaskInfo task : entry.getValue()) {
-                    builder.startObject(task.getTaskId().toString());
+                    builder.field(task.getTaskId().toString());
```
Could you explain the reason for this change? It seems to be against the common pattern that I have seen pretty much everywhere else in the code.

nik9000 replied:
This kind of got pulled along. ToXContent doesn't specify either way but we have lots of stuff that acts like it does spit out a whole object. `Strings#toString`, for example.

Anyway, I hit some broken thing with XContent and started pulling on it to "fix" it the way I was used to seeing other things and it sort of expanded into a lot of changes. I can revert it and break it and deal with it later if you like.

Part of the reason I like it this was is that it means that ToXContent has symmetry with the parser. You can parse exactly what it spits out with what it came from.

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/get/GetTaskRequest.java:
```diff
@@ -0,0 +1,82 @@
+/**
+ * A request to get node tasks
+ */
+public class GetTaskRequest extends BaseTasksRequest<GetTaskRequest> {
```
This might be really confusing. We don't support any parameters that are provided by `BaseTasksRequest<GetTaskRequest>`  so the only reason to use it would be to continue using TransportTasksAction which is overkill here anyway since we know the node that we need to get the task from. So, we can base this directly on TransportNodesAction instead.

nik9000 replied:
Sure! I'm glad I didn't do that before getting your review because I wouldn't have gone the "base it on TransportNodesAction direction - but I think it is a good one. I'll investigate.

imotov commented:
I think it's almost there. I think we just need to finish separation between list tasks and get task by rebasing GetTask on NodeAction instead of TasksAction or by adding a specialized TaskAction and then we should be all set.

nik9000 commented:
@imotov I switched TransportGetTaskAction to extending HandledTransportAction and that cleaned the code up quite a bit! No more null support in GetTaskResponse, no more hijacking, etc.

I'm not experienced with actually building these sorts of transport actions, so please have a look and let me know if it is funky.

I didn't want to extend NodesTransportAction because that wants to work with multiple nodes and wants my request to extends BaseNodesRequest, neither of which make much sense. Instead I just, sort of, bounce execution of the action over to the correct node.

clintongormley commented:
Sorry for the late and somewhat off topic comment, but I think we should call the index `.task_status` or `.tasks` rather than `.results`, which is just too generic.

imotov commented:
`.task_results`?

nik9000 commented:
> .task_results?

What if we want to snapshot running tasks into it? I kind of think we will soon. I prefer `.tasks` if so.

imotov commented:
Yeah, `.tasks` sounds good to me.

nik9000 commented:
> Yeah, .tasks sounds good to me.

Should I make that change as part of this or should we do it after?

imotov commented:
Sure, why not?

nik9000 commented:
> Sure, why not?

Cool. It is time for me to switch branches anyway, so I'll do it now.

nik9000 commented:
@imotov renamed the index, the type, and a few of the things who's names no longer matched. Have a look when you are ready. I tried to make the commit messages useful.

imotov commented on core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/get/TransportGetTaskAction.java:
```diff
@@ -0,0 +1,203 @@
+    private final Client client;
+
+    @Inject
+    public TransportGetTaskAction(Settings settings, ClusterName clusterName, ThreadPool threadPool, TransportService transportService,
```
clusterName is not needed here.

imotov commented on core/src/main/java/org/elasticsearch/tasks/TaskPersistenceService.java:
```diff
@@ -19,36 +19,43 @@
+public class TaskPersistenceService extends AbstractComponent {
 
-    public static final String TASK_RESULT_INDEX = ".results";
+    public static final String TASK_RESULT_INDEX = ".tasks";
```
Shouldn't we rename these constants as well then?

imotov commented:
Simpler and cleaner now. Left a couple of really minor comments. LGTM.

nik9000 commented:
@imotov I squashed and rebased this one to get it ready to merge and found an issue. The wait_for_completion code in the get request was sometimes running in the *requesting* thread. So it'd make the tests timeout and fail. Sometimes. I've moved it to the generic thread pool and the test is stable again.

imotov commented:
LGTM

nik9000 commented:
Thanks for all the reviews and design help @imotov !
